File: src/main/kotlin/checkIfTwoStringArraysAreEquivalent.kt
//       date: 2022-10-25
//      title: Check If Two String Arrays are Equivalent
//        url: https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent/
// difficulty: easy
//------------------------------------------------------------------------------

fun arrayStringsAreEqual(word1: Array<String>, word2: Array<String>): Boolean =
    word1.joinToString("") == word2.joinToString("")

fun arrayStringsAreEqualSeq(word1: Array<String>, word2: Array<String>): Boolean {
    if (word1.sumOf { it.length } != word2.sumOf { it.length }) {
        return false
    }
    val seq1 = word1.asSequence().flatMap { it.asSequence() }
    val seq2 = word2.asSequence().flatMap { it.asSequence() }
    return seq1.zip(seq2).all { it.first == it.second }
}

File: haskell/CheckIfTwoStringArraysAreEquivalent.hs
--       date: 2022-10-25
--      title: Check If Two String Arrays are Equivalent
--        url: https://leetcode.com/problems/check-if-two-string-arrays-are-equivalent/
-- difficulty: easy
--------------------------------------------------------------------------------
module CheckIfTwoStringArraysAreEquivalent
  ( arrayStringsAreEqual
  , tests
  ) where

import           Test.HUnit

arrayStringsAreEqual :: [String] -> [String] -> Bool
arrayStringsAreEqual xs ys = concat xs == concat ys
 
--------------------------------------------------------------------------------
-- Tests
tests :: Test
tests =
  TestList
    [ TestCase $ assertEqual "" True $ arrayStringsAreEqual ["ab", "c"] ["a", "bc"]
    , TestCase $ assertEqual "" False $ arrayStringsAreEqual ["a", "cb"] ["ab", "c"]
    , TestCase $ assertEqual "" True $ arrayStringsAreEqual ["abc", "d", "defg"] ["abcddefg"]
    , TestCase $ assertEqual "" False $ arrayStringsAreEqual ["abc","d","defg"] ["abcddef"]
    ]
